beets/util/__init__.py

```diff
@@ -474,32 +475,45 @@ def copy(path, dest, replace=False):
                               traceback.format_exc())
 
 
-def move(path, dest, replace=False):
+def move(path, dest, replace=False, force=False):
     """Rename a file. `dest` may not be a directory. If `dest` already
     exists, raises an OSError unless `replace` is True. Has no effect if
-    `path` is the same as `dest`. If the paths are on different
-    filesystems (or the rename otherwise fails), a copy is attempted
-    instead, in which case metadata will *not* be preserved. Paths are
-    translated to system paths.
+    `path` is the same as `dest` unless `force` is True. If the paths are
+    on different filesystems (or the rename otherwise fails), a copy
+    is attempted instead, in which case metadata will *not* be preserved.
+    Paths are translated to system paths.
     """
-    if samefile(path, dest):
+    if os.path.isdir(path):
+        raise FilesystemError(u'source is directory', 'move', (path, dest))
+    if samefile(path, dest) and not force:
         return
     path = syspath(path)
     dest = syspath(dest)
-    if os.path.exists(dest) and not replace:
+    if os.path.exists(dest) and not replace and not force:
         raise FilesystemError(u'file exists', 'rename', (path, dest))
 
     # First, try renaming the file.
     try:
-        os.rename(path, dest)
+        os.replace(path, dest)
     except OSError:
         # Otherwise, copy and delete the original.
+        if os.path.isdir(dest):
+            dest = os.path.join(dest, os.path.basename(path))
+        tmp = tempfile.mktemp(suffix='.beets',
+                              prefix=as_string(b'.' + os.path.basename(dest)),
+                              dir=as_string(os.path.dirname(dest)))
```

Comment:
`as_string` is not supposed to be used for paths (see the docstring), in fact I'm pretty sure that it raises an exception for non-utf8 file names (maybe add a test for `beets.util.move()` with non-utf8 filenames?). If this conversion cannot be avoided, use `py3_path` (-> docstring). Maybe this would directly work with bytestrings if using a bytestring suffix `b'.beets'`?

Reply:
@wisp3rwind the root cause of this`as_string` is `tempfile.mktem`. It try to concate `prefix` and `suffix` with string which is obtained via `tempfile._get_candidate_names`.

Anyway, I love an idea about `py3_path`